### ISSUE TITLE
asix.xif leads to random crashes/freezes

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#273]
I started noticing this when I was putting more stuff info mint.cnf, namely:
- usb `loader.prg`
- gluestik.prg
- `ifconfig` & `route` commands

When I put those three lines for the first time, it would crash all the time. Then I (out of despair) tried to replace the `ifconfig` and `route` commands with `/bin/sh /bin/network.sh` (where `network.sh` would contain basically the same thing) and crashes stopped.

Then I tried to compile a 030 version of asix.xif and I got the crashes back again.

Sometimes booting would succeed but when xaloader is starting xaaes, it crashes on a null pointer.

And sometimes it's really random... for instance, this is when `gluestik.prg` is about to be executed:

![PXL_20220730_222930575](https://user-images.githubusercontent.com/3062260/182002422-a3c1b578-ec78-40ce-a765-aa451d7ff545.jpg)

(the crash/assert is always the same)

Another interesting thing is that _if_ it succeeds, then the setup is rock solid. It really looks like some kind of race condition during init. But not module init itself -- I can for instance remove all three lines from mint.cnf, start `loader.prg`, `gluestik.prg` and `network.sh` from deskop and it would sometimes crash even there. So it must be somehow related to assigning the network interface or something like that.

**Comment by mikrosk** [comment]
Hmmm... looking into asix.c has revealed some interesting information:
- the assert comes from kernel's `kfree` function and `asixpoll` does indeed calls kfree() in `asix_poll_thread` (which is in turn awakened from `asix_poll`)
- `asix_poll` is initialised from `asix_recv` to be called every 50ms on 68000 and every 10ms on 68020+ (i.e. the more frequent crashes with my 030 build make sense!)

**Comment by czietz** [comment]
That is a very interesting observation. Yesterday, while working on my
(not yet published) driver based on asix.c, I was seeing the exact same
thing: Random crashes, sometimes when the network was brought up by
ifconfig and route, sometimes right after it during xaloader. Same as
with you: _If_ it boots successfully, it runs totally solid.

I thought this had to be a bug in my driver and unsuccessfully reviewed
my changes again and again. But since you see the crashes with the asix
driver, too, this gives me further ideas.

**Comment by czietz** [comment]
Okay, my current theory: What if `asix_recv()` is called via the `asix_poll_thread` while the network stack is still in the process of bringing the interface up? As you say: a race condition.

What seems to fix the problem for me, is thus:
a) Putting the poll thread to sleep immediately:
```
void
asix_poll_thread(void *e)
{
	struct eth_device *eth = (struct eth_device *)e;
	struct ueth_data *dev = (struct ueth_data *)eth->data;

	sleep(WAIT_Q, (long)&asix_poll_thread);

	[...]
```
b) Only waking it up for the first time with some delay at the end of `asix_init()`:
```
	addroottimeout(250, asix_poll, 0);
```

PS: I did again test with my driver. Would be interesting to see if you can confirm this with the original asix driver, too.

**Comment by mikrosk** [comment]
Unfortunately, I can't confirm your findings.

In the meantime I have found quite good test case: replacing this kmalloc: https://github.com/freemint/freemint/blob/master/sys/sockets/xif/asix.c#L617-L621 with `static unsigned char recv_buf[AX_RX_URB_SIZE];`, this triggers the assert every time (only confirming that it is a timing-sensitive issue).

So after applying the two changes I can say that the assert is still there (I always confirm that the setup is working by booting the "clean" version of asix.xif, compiled with `-m68030` -- this flavour looks like most stable for me) . First I misread that I should try the two lines separately, so for what it's worth, that lead to crashes as well. :)

**Comment by czietz** [comment]
Interesting. For me, the changes (both of them at the same time) seem to have fixed it. Perhaps just by chance and it's something entirely different? Or maybe a longer delay is required in your case? The 250 ms were just an arbitrary guess.

**Comment by mikrosk** [comment]
I'll try to play with it tonight. Maybe you can try my "repro", i.e. the static buffer + -m68030 to see what happens?

**Comment by mikrosk** [comment]
No, nothing. Tried 250, 500, 750, 1000, even 1 just for fun, always the same effect (or lack of any). This is my change, I assume there's not much I could do wrong: https://github.com/mikrosk/freemint/commit/ab0ef0900470c6886f8c9c10482ce145d5bd48a8

**Comment by mikrosk** [comment]
Scratch that, I really am an idiot. So another round of testing...

**Comment by mikrosk** [comment]
Ok, so one more time.

This is my current set of changes: https://github.com/mikrosk/freemint/commit/6275226012e650482f25c945137d4dfebbdf43bc. However all my reports with the static buffer are worthless because only an hour ago I realised that I missed this kfree() call: https://github.com/mikrosk/freemint/commit/6275226012e650482f25c945137d4dfebbdf43bc#diff-45d88974d297f8d589204c9d9e337cb325349b05172180b8f2193ca1399d0724L709 and therefore I was getting exactly the same asserts, just from different thread (and for a different reason - my oversight).

However after this, erm, correction all my 030 builds (with the two lines and/or with the static buffer or the original, just with -m68030) simply work and I'm unable to crash them no matter what I do.

So I switched my attention to -m68000 builds, to make things simpler I was testing just the original asix.xif and a custom one with the two lines added (i.e. without the static buffer).

And all I can say is that... I'm clueless. I've learnt only one hard lesson, it's very important to power cycle between tests (or at least CTRL+ALT+SHIFT+DEL to clear memory incl. TT RAM) because the nature of that bug is that it reads some random memory and it seems that sometimes that helps triggering the bug more often while sometimes it does the opposite.

So in the end... I was able to see crashes with the original asix.xif as well as with the fixed one. There wasn't anything specific or repeatable, any kind of pattern. Sometimes I would see the fixed one crashing 5 times in a row (between power cycles!!!), then I would switch to the original asix.xif, boot, see it crash, power cycle and not crashing for another 5 power cycles.

The only common denominator was that I wouldn't see the assert again, it would always crash after xaloader execution. It's totally possible that the mysterious asserts came from the fact that there was some garbage in memory and after new boot the driver would again read some illegal memory but this time instead of accidental zeroes it would pick up some random numbers.

So @czietz be careful with your observations, it really seems to be working by chance.

**Comment by czietz** [comment]
> And all I can say is that... I'm clueless. I've learnt only one hard lesson, it's very important to power cycle between tests (or at least CTRL+ALT+SHIFT+DEL to clear memory incl. TT RAM) because the nature of that bug is that it reads some random memory and it seems that sometimes that helps triggering the bug more often while sometimes it does the opposite.

That matches what I observed. _If_ it worked, it would still work after a reset. Only after a power cycle it could at random switch to the non-working state. However, I have done dozens of power cycles during my development today, without a single crash ever since I made the change described in https://github.com/freemint/freemint/issues/273#issuecomment-1200374665.

Since our observations seem to suggest some random memory contents, there is one thing you could investigate: Whether the BSS of .xif modules is properly dimensioned and zeroed when they are loaded.

**Comment by mikrosk** [comment]
> there is one thing you could investigate: Whether the BSS of .xif modules is properly dimensioned and zeroed when they are loaded.

I think this should be covered here: https://github.com/freemint/freemint/blob/master/sys/module.c#L243-L257. What I think is happening is that [dev->pusb_dev](https://github.com/freemint/freemint/blob/master/sys/sockets/xif/asix.c#L966) and/or [dev->dev_priv](https://github.com/freemint/freemint/blob/master/sys/sockets/xif/asix.c#L972) are dangling pointers. I have some ideas what to check and how to debug it, hopefully I'll get to it today.

**Comment by mikrosk** [comment]
Even after two evenings spent on this I'm no closer to at least some clue. I've set up quite good debugging environment (serial port output on another computer) but due to the time sensitive nature of the bug, enabling debugging didn't exactly help, just made the bug even harder to reproduce (or, if I enable too much output, crippling the kernel & xaaes to a hardly usable state).

I was able to verify that all pointers are good (at least that nobody has changed them from their initial values). I also verified code which allocates and releases their memory, confirmed by debug outputs, nothing suspicious is going on there in the thread body (I saw a freeze but all pointers were still in a healthy state).

In one stroke of luck I was able to have not-so-spamming config yet with the thread heartbeat enabled (so I could see every failed attempt to receive something -- the thread was sleeping in that exact moment) as well as outputs in asix_output(). And really, I could see it stopping there. I had an output just before this line: https://github.com/freemint/freemint/blob/master/sys/sockets/xif/asix.c#L740 ... it was printed as part of `ifconfig` process and then... nothing. The thread has stopped, booting stopped.

My guesses currently are:
- `asix_poll_thread`'s asix_recv() somehow collides with the rest of the usb stack (i.e. it tries to receive when a send is in progress and messes something up ... however see the paragraph above, it seemed that the thread was sleeping when the freeze occured)
- `asix_poll_thread`'s asix_recv() somehow collides with the whole system, i.e. when an executable or kernel module (xaaes) is loaded, maybe there is something bad happening if the thread is currently active. The crash/freeze never happens when nothing is happening, there always must be something to be loaded. That would also explain why @czietz's fix seems to be working - it basically offsets thread activity until ifconfig/xaloader has been executed.
- interestingly, `asix_send` seems to be sometimes called from `ifconfig` process and sometimes from `asixpoll` (could be the debug printer, though, not sure).

**Comment by czietz** [comment]
>   * |asix_poll_thread|'s asix_recv() somehow collides with the rest of
>     the usb stack (i.e. it tries to receive when a send is in progress
>     and messes something up ... however see the paragraph above, it
>     seemed that the thread was sleeping when the freeze occured)


Sadly, I cannot assist with any actual debugging right now due to lack
of time. But there was one observation I made last weekend, that seemed
to indicate that that handing off of the packets in asix_recv() is the
issue. When I commented out the calls to bpf_input (no idea if that is
even used) and if_input I did not see the crash anymore. Of course, MiNT
did not see any received packets anymore, either, due to this change.

As you know the bug lacks some reproducibility; so, my observation needs
to be taken with the proverbial grain of salt.

**Comment by mikrosk** [comment]
No worries Christian, that was just my personal grievance. :) However, just a few minutes after posting it, I got really lucky. I had inet4.xdd + usb.km + eth.udd in debug mode (asix.xif was the original) and I was able to once boot into a frozen state and just after that into a workingstate! I'm attaching full logs.

Left - booted into XaAES, right - stuck on `route` command:

![image](https://user-images.githubusercontent.com/3062260/182464424-17bab259-d543-47b1-b039-5ff3dfa6983e.png)

[freezes.txt](https://github.com/freemint/freemint/files/9245304/freezes.txt)
[works.txt](https://github.com/freemint/freemint/files/9245305/works.txt)

**Comment by czietz** [comment]
What is apparent: The sequence of steps is different. In the working case the default route is added later. This might simply be due to slightly different scheduling of the individual processes and threads. But as to why it freezes, I would have to think about it much longer.

**Comment by mikrosk** [comment]
Indeed. I also took a look at this afterwards (I was exhausted that day) and yes, the only obvious thing is that if "something" happens before ifconfig/route command, then it's fine while if it happens after/during those commands, then it leads to freeze.

Very good catch with the routing address, as you can see, my (statically) assigned IP is 192.168.1.2 so there is no way that 74.136.102.240 could be pulled from somewhere (that IP is assigned in mint.cnf with the `ifconfig` command).

Since debug output seem to weaker the repro, I'll try to focus solely on this fact, i.e. output just the IP and see if corrupt IP correlates with the freezes.

**Comment by czietz** [comment]
Guess what: I've commented out some code in my driver (e.g., the actual USB transfer to _receive_ packets in the ..._recv() function), making it faster (and unable to receive data, of course). Now I'm reliably getting the exact same assert that you were seeing in your [first post](https://github.com/freemint/freemint/issues/273#issue-1323313236). Argh, what the ...??

**Comment by mikrosk** [comment]
I'm not sure whether to celebrate or share your pain. :-) Make sure you don't have any orphaned kfree() in your code (so you won't get distracted the same way as I did when I thought that I found a stable repro) but otherwise it would be nice to have a stable repro.

Btw, I'm unable to again get that assert at all. Always frozen ifconfig/route/xaloader these days.

**Comment by mikrosk** [comment]
> Btw, I'm unable to again get that assert at all. Always frozen ifconfig/route/xaloader these days.

And even that is an understatement as it seems. After the lucky debug dump two days ago, I'm unable to crash it _at all_. I've tried every thing I'd tried before -- 030 version, static buffer vs kmalloc, disabling all debug output entirely, reverting to original binaries. So I really hope that you've succeeded with finding something more reproducible (I even tried to disable the whole code in `asix_recv` except the usb bulk message and it didn't make difference either).

Btw, one thing has caught my attention but since I'm unable to experiment with it... in `asix_recv`, we call `buf_alloc` with `BUF_ATOMIC`. This driver is one of only two which combine `kthread_create` with `buf_alloc`, the other one is serial + ppp.

If you take a look what is [happening](https://github.com/freemint/freemint/blob/master/sys/sockets/buf.c#L311) in `buf_alloc`, this parameter makes the allocator repeatedly calling `addroottimeout`, same as we do when handling `asix_recv`. So what if these two threads somehow collide? Maybe a nice test would be to change it to `BUF_NORMAL` , if it changes something.

Another idea I had was to change the thread to an interrupt, same as other drivers do, just to see what happens.

But all of this begs for a stable repro to experiment with.

**Comment by czietz** [comment]
I've also thought about kernel threads; albeit in a different direction.

Where do kthreads get their stack from? Is it ensured that each thread
gets its separate and sufficiently sized stack? If you have concurrent
threads with overlapping stacks, you can have the weirdest corruption...

**Comment by mikrosk** [comment]
Hmm, thinking along the lines of stack corruption: just [recently](https://github.com/freemint/freemint/commit/48897708cc409a78689d326a8e38416e448f7120) I was investigating an issue related to stack corruption (so yes, I was looking for a place of stack initialisation as well). If you take a look at the hint we have (completely bogus IP addr here: https://github.com/freemint/freemint/blob/master/sys/sockets/inet4/route.c#L72) at first it could look that yes, we might be indeed feeding it from a corrupted stack. However if that would be the case, it would be quite unlikely to survive so long without a crash.

So what about other callers? If you search for `route_get`, you'll see that awfully lot of times `struct BUF*` is involved and also there are sometimes (de)allocations with `BUF_NORMAL` happening.

But as I said, to have something to confirm/reject these leads I'd need a stable repro, or at least _a_ repro. :-/

**Comment by czietz** [comment]
**I think I might have found it!**

After seriously considering that the bug was mocking me, because every day I got a different crash and on some days I could not even reproduce the working state, I decided to systematically check for uninitialized memory. As discussed before, something depending on essentially random memory contents explains the problems with reproducibility.

Anyway, during my review, I found this:
https://github.com/freemint/freemint/blob/ab9c65b589e11d9f6c796974b7c2c99d1e14e4fd/sys/usb/src.km/udd/eth/usb_ether.c#L296

usb_eth is where basically all information about USB network adapters is being kept. It is a pointer, though:
https://github.com/freemint/freemint/blob/ab9c65b589e11d9f6c796974b7c2c99d1e14e4fd/sys/usb/src.km/udd/eth/usb_ether.c#L101
... and, hence, `sizeof(usb_eth)` evaluates to 4 (32 bit). `usb_eth` is not properly zeroed. I think the author intended to write:
```
memset(usb_eth, 0, sizeof(struct ueth_data) * usbNetAPI->numDevices);
```

After replacing the line, I did not observe any weird issues (crashes, freezes, error messages) anymore; neither with my own USB network device, nor with the Asix adapter!

Of course, due to the reproducibility crisis with this bug, I can only be certain after observing for a lot of power cycles. However, I think we can agree that the memset() as it currently is is wrong. Perhaps you could test this change on your side?

**Comment by mikrosk** [comment]
That is indeed a great find! I like your systematic and patient approach much more than mine chaotic guessing. :-)

I've given it a go on my TT+Lightning+ASIX as well as CT60+NetUSBee+ASIX. The only suspicious thing was that I did see a freeze on the TT while starting xaaes until I pushed the reset button (and eventually power cycled), i.e. soft (CTRL+ALT+DEL) reset still didn't seem to get rid of some memory corruption entirely. But that could be also something totally unrelated.

Feel free to push your fix, that was a bug without any doubt. In the meantime I pushed two of my helper commits for `asix.c`.

**Comment by vinriviere** [comment]
> I think I might have found it!

Hats off, @czietz 😃

**Comment by th-otto** [comment]
Nice find. I just wonder why we don't get a warning with newer versions of gcc, which normally detects this?

**Comment by czietz** [comment]
> Nice find. I just wonder why we don't get a warning with newer versions of gcc, which normally detects this?

I assume this is because for kernel modules (like eth.udd) memset() is replaced by a define:
https://github.com/freemint/freemint/blob/2f027a2dbbc306b1bb628397307f67c69295d840/sys/libkern/kernel_module.h#L1086
... and, therefore, gcc does not recognize the call as standard memset(), anymore.

**Comment by th-otto** [comment]
I don't see any warnings in the explorer, but of course that explains it.

Would be nice if we could find a way to tell gcc that this is really memset (similar for other standard functions). Who knows what other warnings may pop up from this. Although of couse, in this case it was not just a ```sizeof(usb_eth)``` vs ```sizeof(*usb_eth)```

**Comment by mikrosk** [comment]
Just an update on this: this bug is not related to asix.xif. I got the exact same crash on a totally different setup:
- different Falcon, clean disk
- latest (26.10.2023) kernel (which is, since some time, compiled with `-fno-strict-aliasing`
- Ethernec driver from Ozk (https://assemsoft.atari.org/ne2k/index.html, the CT2 version)
- CT60 + NetUSBee
- USB loader.prg not loaded at all, let alone asix.xif
- the crash happened, again, after executing the `ifconfig` + `route` combo, this time from `exec u:\bin\bash u:\c\mint\network.sh`, so basically the opposite of what I had in the first post

**Comment by skaftetryne** [comment]
Similar issue described here: https://sourceforge.net/p/freemint/mailman/freemint-discuss/thread/76345003e00aa9b4622c3a0074cfdcf19afb06e9.camel%40online.no/#msg59119965

**Comment by mikrosk** [comment]
@skaftetryne you mentioned a CTRL+ALT+DEL issue in XaAES in your email back then. As you are one of the few SuperVidel+FreeMiNT users I know of, haven't you noticed a pretty stable pattern if you CTRL+ALT+DEL with svethlana.xif and `ifconfig` being resident? To me there's always a crash within first 2-3 seconds after the "Initializing SD-RAM"  CT60 TOS boot message (and during counting free memory).

Because I have found the culprit so I'm wondering whether you could test it afterwards, too. Assuming that you are seeing the same problem as I do.

**Comment by skaftetryne** [comment]
I'm afraid the only networking hardware I have for my Falcon is the Hydra (with USB, which came as a surprise to me!) clone I bought from you a while ago. I've not had any reboot-issues since you fixed this bug last year.
